File: src/matlab/+jterator/+api/+io/read_input_args.m
%% Reading input arguments from HDF5 file
%% using the location specified in "handles".

function input_args = read_input_args(handles)

    import jterator.api.h5.*;
    import jterator.api.etc.*;
    

    hdf5_filename = handles.hdf5_filename;

    input_args = struct();
    keys = fieldnames(handles.input_keys);
    for i = 1:length(keys)
        key = keys{i};
        field = handles.input_keys.(key);

        if isfield(field, 'hdf5_location')
            input_args.(key).variable = h5varget(hdf5_filename, field.hdf5_location);
            fprintf(sprintf('jt -- %s: loaded dataset ''%s'' from HDF5 group: "%s"\n', ...
                    mfilename, key, field.hdf5_location))
        elseif isfield(field, 'parameter')
            input_args.(key).variable = field.parameter;

            % temporary hack: this could also be fixed in ReadYaml.m
            % (more specifically in subfunction makematrices.m)
            if iscell(field.parameter)
                if all(cellfun(@isnumeric, field.parameter))
                    input_args.(key).variable = cell2mat(field.parameter);
                end
            end

            if ischar(field.parameter)
                fprintf(sprintf('jt -- %s: parameter ''%s'': "%s"\n', ...
                        mfilename, key, input_args.(key).variable))
            else
                fprintf(sprintf('jt -- %s: parameter ''%s'': %s\n', ...
                        mfilename, key, vec2str(input_args.(key).variable)))
            end
        else
            error('Possible variable keys are ''hdf5_location'' or ''parameter''');
        end 

        if isfield(field, 'class')
            input_args.(key).class = {field.class};
        else
            input_args.(key).class = {};
        end

        if isfield(field, 'attributes')
            input_args.(key).attributes = {field.attributes};
        else
            input_args.(key).attributes = {};
        end   

    end
    
end

File: src/r/jterator/R/read_input_args.R
read_input_args <-
function(handles) {

    mfilename <- "read_input_args"

    hdf5_filename <- handles$hdf5_filename

    input_args <- list()
    for (key in names(handles$input_keys)) {
      field <- handles$input_keys[[key]]
      input_args[[key]] <- list()

      if ("hdf5_location" %in% names(field)) {
        input_args[[key]]$variable <- h5read(hdf5_filename, field$hdf5_location)
        cat(sprintf("jt -- %s: loaded dataset '%s' from HDF5 group: \"%s\"\n",
                mfilename, key, field$hdf5_location))
      }
      else if ("parameter" %in% names(field)) {
        input_args[[key]]$variable <- field$parameter
        cat(sprintf("jt -- %s: parameter '%s': \"%s\"\n",
                    mfilename, key, paste(field$parameter, collapse=",")))
      }
      else {
        stop("Possible variable keys are \'hdf5_location\' or \'parameter\'")
      }

      if ("class" %in% names(field)) {
        input_args[[key]]$class <- field$class
      }
      
      if ("attributes" %in% names(field)) {
        input_args[[key]]$attributes <- field$attributes
      }
    }

    return(input_args)
}
